Secure, user-friendly and feature-rich authorization method

rollersteaam opened the pull request:
This feature was based off the challenges discussed in discussion #276 

SecureAuthorizationCodeAuth is a rework of how AuthorizationCodeAuth operates. It was developed for the use case scenario where the developer has created their own *exchange server** to perform all authorization requests with Spotify.

After trying to figure out what the use of a 'backend server' would mean, I developed a solution shown in the documentation addition below using 2 php files hosted on a php website server that would perform: 'get authorization code', 'get access token and refresh token', 'get refreshed access token from refresh token'.

Inheriting from SpotifyAuthServer<T> provided by 3.0.0, I created SecureAuthorizationCodeAuth so I could define the required functions to interact with a user's exchange server. I extended the functionality of it compared to other auth methods by adding an access token expired event, and auth success and failure events and the ability for devs to auto refresh their access to Spotify. Additionally, I implemented a 'silent refresh', which allows refreshing of access without requiring opening the user's browser. The only time that is mandatory is for program initialization essentially - the rest is magic, the access expiry times may as well not exist. It provides the illusion that it is permanent because it works so well. On top of this, the client secret is never exposed in the process, and the setup is as easy as copy and pasting the two php files into your website and inputting the necessary client id, secret, redirect, etc.

Personally, this solved three main problems of mine that existed with SpotifyAPI-NET so far: the need to open the browser to refresh authorization, the frequency at which refreshing authorization had to happen (Implicit Grant) and having to monitor the expiry time myself by setting up my own timer.

I have already detailed very descriptive documentation and comments across the two new files as well as adding the necessary page in the SpotifyAPI-NET wiki.

OmiCron07 commented:
@rollersteaam 

In the class `SecureAuthorizationCodeAuth`, you set a timer with `Interval = token.ExpiresIn`, but `Interval` is in milliseconds and `token.ExpiresIn` in seconds. You have to multiply by 1000 to have 60 minutes instead of 3.6 seconds.

Another thing, I don't know if that problem is linked with the error above, but sometime in `RefreshAuthAsync`, the method `GetToken` returns null and `SetAccessExpireTimer` will never get called, so if that call fails, the token never gets refreshed. I don't know why sometime the token can't be refreshed, maybe a glitch in my Internet or Spotify server is busy, but it should have mechanism to ensure it retries a given time before completely failing. I got the null token a couple of times yet.

rollersteaam commented:
@OmiCron07 Sorry for the delay, only got a moment to code today. I've now fixed the ExpiresIn issue and I've added auto retrying for tokens. Its default value is 10 retries but you may vary it in SecureWebAPIFactory and SecureAuthorizationCodeAuth as you like. Additionally, there are rare cases where a refresh or exchange attempt may falsely pass as successful when no AccessToken property value is present in a GetToken request. I've improved token validation by adding for checks for this.

OmiCron07 commented:
@rollersteaam Hey, no problem! I saw the commits, amazing work! I will pull the changes later today to test them out. Thank you!

rollersteaam commented:
@OmiCron07 Thanks! It's probably worth noting I've merged all of the new things that were on the original branch (https://github.com/JohnnyCrazy/SpotifyAPI-NET) which should resolve the issue you had with currently_playing_type, which means my branch has "caught up" now

Lycrosa commented:
Hey, 

do you have a eta for when this pullrequest gets through? Would be really interested to use this.

Kind Regards

Pascal

rollersteaam commented:
> Hey,
> 
> do you have a eta for when this pullrequest gets through? Would be really interested to use this.
> 
> Kind Regards
> 
> Pascal

https://github.com/JohnnyCrazy/SpotifyAPI-NET/issues/298

Here JohnnyCrazy, the author, mentions he needs some more time reviewing it. It may be a while before this gets accepted but you can always choose to pull this branch instead.

Lycrosa commented:
Allright thank you for answering!

camcam2015 commented:
@rollersteaam 

Hi, I just had a quick and probably really basic question for you so apologies in advance if the answer is extremely obvious! I'm new to working with APIs but it's been fun learning how to use them.

I see in the docs how you can use `webApiFactory` as a streamlined and safe way to get the auth and token for Spotify, yet when I try to use it the class itself isn't found. I'm using the `SpotifyAPI.Web`
and `SpotifyAPI.Web.Auth` namespaces, is there something I'm missing here or skipped over?

JohnnyCrazy commented:
@camcam2015 The docs are not updated yet and this PR is not in the official nuget yet.

I planned the review/merge and doc update for tomorrow, so I hope it won't take any longer, sorry for the delay guys

rollersteaam commented:
@camcam2015 Hiya, we've all been there when using libraries for the first time.

To use what I've made you would need to pull this branch, open the project file and then build the project yourself. The DLLs that produces would need to replace the ones where JohnnyCrazy's original SpotifyAPI DLLs are, which should be in the "packages" folder inside your project under folders such as "SpotifyAPI.Web" and "SpotifyAPI.Web.Auth".

However to save you some time I did the building of the DLLs for you:

[Click here to download the DLLs for my pull request's version](https://dl.dropboxusercontent.com/s/e2mvu0j6a2ikw8e/rollersteaam_SpotifyAPI-Net%20binaries.zip)

Now you just need to place those folders in the "packages" folder inside your project and replace all. You should be wary that this is 1 commit behind the main branch, JohnnyCrazy's, and it might already be integrated soon so you can do this just to try it out for the moment I guess. Happy coding

@JohnnyCrazy No worries, thanks for getting around to it

JohnnyCrazy commented on SpotifyAPI.Docs/docs/SpotifyWebAPI/auth.md:
````diff
@@ -100,6 +102,277 @@ static void auth_OnResponseReceivedEvent(Token token, string state, string error
+auth.OpenBrowser();
+```
+
+## Creating the Exchange Server
````
I would like to abstract this a little bit more in the docs and keep the PHP Script in a repo of your choice. (Also this should be ### AFAIK)

Sth like the following:
```
### Token Swap Endpoint

To use this method, an external HTTP Server needs to be able to supply the following HTTP Endpoints to our application:
`/swap` - Swaps out an `authorization_code` with an `access_token` and `refresh_token` - The following parameters are required in the JSON POST Body: ...
`/refresh` - Refreshes an `access_token` - The following parameters are required in the JSON POST Body: ...

The following open-source token swap endpoints can be used:
rollersteaam/spotify-token-swap-php
simontaen/SpotifyTokenSwap (note: would be cool if we could support 2 servers at once!)
```
I also don't think we need the redirect functionality on the server, every client should be able to build this URL themselves

rollersteaam replied:
I have refactored the Creating the Exchange Server doc entry for your Token Swap Endpoint entry in this commit https://github.com/rollersteaam/SpotifyAPI-NET/commit/321724830f604d8fcb9275aa6136ff3818c4f864.

But I'm not sure if putting the redirect functionality on the server should be avoided or not, so until that's decided the PR won't be able to move forward yet, as if it shouldn't be avoided, some more information will need to be added to the docs (which I'll do as necessary).

JohnnyCrazy replied:
It should be fine in this state, will check again tho

JohnnyCrazy replied:
Regarding this quote:

> However the only reason why I have a redirect for the GET auth code request to Spotify in my php example is that it keeps the client_id and redirect_uri off a program so that anyone decompiling the program can't find that information.

Well, the information is in the login URL of spotify anyway, so instead of decompiling, a user just has to run the program. Also, `client_id` and `redirect_uri` can be public. However, the current `ImplicitGrantAuth` solution requires that you use localhost as the redirect URI, which allows everyone to use this "domain" and create `ImplicitGrantAuth` as well. This is fixed by your auth solution.

(The only drawback of this solution is, that every user can simply request an unlimited amount of tokens by hitting the endpoint often. However, this is normal and can't be avoided when having an anonymous client side app. So in the end, this solution is the best we can currently support 👍 )

JohnnyCrazy commented on SpotifyAPI.Web.Auth/SecureAuthorizationCodeAuth.cs:
```diff
@@ -0,0 +1,210 @@
+    /// It's recommended that you use <see cref="SecureWebAPIFactory"/> if you would like to use the SecureAuthorizationCodeAuth method.
+    /// </para>
+    /// <para>
+    /// If you are unsure what the exchange server is or how to make one, see <see href="https://johnnycrazy.github.io/SpotifyAPI-NET/SpotifyWebAPI/auth/#secureauthorizationcodeauth"/> (the SecureAuthorizationCodeAuth page on the SpotifyAPI-NET documentation).
```
Don't hardcode any links - These can change quite fast and lead to 404s

JohnnyCrazy commented on SpotifyAPI.Web.Auth/SecureAuthorizationCodeAuth.cs:
```diff
@@ -0,0 +1,210 @@
+        /// A version of <see cref="AuthorizationCodeAuth"/> that does not store your client secret, client ID or redirect URI, enforcing a secure authorization flow. Requires an exchange server that will return the authorization code to its callback server via GET request.
+        /// </para>
+        /// <para>
+        /// If you are unsure what the exchange server is or how to make one, see <see href="https://johnnycrazy.github.io/SpotifyAPI-NET/SpotifyWebAPI/auth/#secureauthorizationcodeauth"/> (the SecureAuthorizationCodeAuth page on the SpotifyAPI-NET documentation).
```
Don't hardcode any links - These can change quite fast and lead to 404s

JohnnyCrazy commented on SpotifyAPI.Web.Auth/SecureWebAPIFactory.cs:
```diff
@@ -0,0 +1,270 @@
+    /// <summary>
+    /// Returns a SpotifyWebAPI using the SecureAuthorizationCodeAuth process.
+    /// <para/>
+    /// For more information, see <see href="https://johnnycrazy.github.io/SpotifyAPI-NET/SpotifyWebAPI/auth/#secureauthorizationcodeauth"/> (the SecureAuthorizationCodeAuth page on the SpotifyAPI-Net documentation).
```
Don't hardcode any links - These can change quite fast and lead to 404s

JohnnyCrazy commented on SpotifyAPI.Web.Auth/SecureWebAPIFactory.cs:
```diff
@@ -0,0 +1,270 @@
+        /// <summary>
+        /// Returns a SpotifyWebAPI using the SecureAuthorizationCodeAuth process. This will not work unless you implement an exchange server.
+        /// <para/>
+        /// For more information, see <see href="https://johnnycrazy.github.io/SpotifyAPI-NET/SpotifyWebAPI/auth/#secureauthorizationcodeauth"/> (the SecureAuthorizationCodeAuth page on the SpotifyAPI-Net documentation).
```
Don't hardcode any links - These can change quite fast and lead to 404s

rollersteaam commented on SpotifyAPI.Docs/docs/SpotifyWebAPI/auth.md:
```diff
@@ -12,13 +12,15 @@ After you created your Application, you will have following important values:
 * [ImplicitGrantAuth](/SpotifyWebAPI/auth#implicitgrantauth) (**Recommended**, no server-side code needed)  
 
-* [AutorizationCodeAuth](/SpotifyWebAPI/auth#autorizationcodeauth) (Not Recommended, Server-side code needed, else it's unsecure)
+* [SecureAuthorizationCodeAuth](/SpotifyWebAPI/auth#secureauthorizationcodeauth) (**Recommended**, server-side code mandatory, most secure method. The necessary code is shown here so you do not have to code it yourself.)
```
This sounds like a great idea, I wasn't sure what to name it as initially anyway so I thought I'd just give a homage to the original local API factory haha. I will change all names to refer to it as such when I can.

JohnnyCrazy commented:
Thank you very much for the updated @rollersteaam , really appreciate it!

I hope I will find time for the merge process (I'm also gonna go ahead and fix remaining issues, if any) and this should be upstream & in the next version soon. Sorry for the delay!

JohnnyCrazy commented:
Ready for merge or do you want to add anything? :) @rollersteaam

rollersteaam commented:
@JohnnyCrazy Absolutely! Ready for merge. From my application's testing I don't think I need to add anything else.

JohnnyCrazy commented:
Alright!

Thanks for this PR, really appreciate the additions :)

rollersteaam commented:
No problem, thanks for the guidance along the way.
